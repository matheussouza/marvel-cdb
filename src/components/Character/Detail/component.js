import React from 'react';

export default ({ character, edit, history }) => {
  return (
    <div>
      <p>
        <i data-testid="description">{character.description || 'Description unavailable.'}</i>
      </p>
      <hr />
      <div>
        <div className="d-flex align-items-center justify-content-between">
          <h5>Extras</h5>
          <div>
            <button
              data-testid="series-btn"
              disabled={ character.series.available === 0 }
              className="btn btn-primary"
              onClick={ () => history.push(`/${character.id}/series`) }
            >
              Series
            </button>
            <button
              data-testid="edit-btn"
              className="btn btn-warning"
              style={{ marginLeft: 10 }}
              onClick={ () => edit() }
            >
              <i className="fa fa-edit" /> Edit
            </button>
          </div>
        </div>
      </div>
      <hr />
    </div>
  )
};
