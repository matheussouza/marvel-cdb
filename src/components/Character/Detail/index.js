import Detail from './component';
import { withRouter } from 'react-router-dom'

export default withRouter(Detail);