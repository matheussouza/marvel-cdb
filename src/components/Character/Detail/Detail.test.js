import React from 'react';
import { render, fireEvent } from '@testing-library/react';

import Detail from './component';

const buildCharacter = (id, description, series) => ({
  id, description, series
});


describe('Character Detail component', () => {
  
  let historyFunctionMock;
  let editFunctionMock;
  beforeEach(() => {
    editFunctionMock = jest.fn();
    historyFunctionMock = jest.fn();
  });
  

  it('should render', () => {
    const char = buildCharacter(1, 'Your friendly neighborhood.', {});
    const subject = render(<Detail character={ char } />);
    expect(subject.container).toMatchSnapshot();
  });

  it('should contains disabled series button with 0 available', () => {
    const char = buildCharacter(1, 'Your friendly neighborhood.', { available: 0 });
    const subject = render(<Detail character={ char } />);
    expect(subject.getByTestId('series-btn')).toHaveProperty('disabled', true);
  });

  it('should contains enabled series button', () => {
    const char = buildCharacter(1, 'Your friendly neighborhood.', {});
    const subject = render(<Detail character={ char } />);
    expect(subject.getByTestId('edit-btn')).toHaveProperty('disabled', false);
  });

  it('should contains default description', () => {
    const char = buildCharacter(1, null, {});
    const subject = render(<Detail character={ char } />);
    expect(subject.getByTestId('description').innerHTML).toMatch('Description unavailable.');
  });

  it('should contains used description', () => {
    const char = buildCharacter(1, 'Your friendly neighborhood.', {});
    const subject = render(<Detail character={ char } />);
    expect(subject.getByTestId('description').innerHTML).toMatch('Your friendly neighborhood.');
  });

  it('should route to series', () => {
    const char = buildCharacter(1, 'Your friendly neighborhood.', {});
    const subject = render(<Detail character={ char } history={{ push: historyFunctionMock }} />);
    expect(historyFunctionMock).toBeCalledTimes(0);
    fireEvent.click(subject.getByTestId('series-btn'));
    expect(historyFunctionMock).toBeCalledTimes(1);
    expect(historyFunctionMock).toBeCalledWith('/1/series');
  });

  it('should call edit function', () => {
    const char = buildCharacter(1, 'Your friendly neighborhood.', {});
    const subject = render(<Detail character={ char } edit={editFunctionMock} />);
    expect(editFunctionMock).toBeCalledTimes(0);
    fireEvent.click(subject.getByTestId('edit-btn'));
    expect(editFunctionMock).toBeCalledTimes(1);
  });

});