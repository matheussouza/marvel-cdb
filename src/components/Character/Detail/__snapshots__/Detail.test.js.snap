// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Character Detail component should render 1`] = `
<div>
  <div>
    <p>
      <i
        data-testid="description"
      >
        Your friendly neighborhood.
      </i>
    </p>
    <hr />
    <div>
      <div
        class="d-flex align-items-center justify-content-between"
      >
        <h5>
          Extras
        </h5>
        <div>
          <button
            class="btn btn-primary"
            data-testid="series-btn"
          >
            Series
          </button>
          <button
            class="btn btn-warning"
            data-testid="edit-btn"
            style="margin-left: 10px;"
          >
            <i
              class="fa fa-edit"
            />
             Edit
          </button>
        </div>
      </div>
    </div>
    <hr />
  </div>
</div>
`;
